File: packages/shared/src/native-status.ts
/** Local observations are not evidence that central OCR, analysis or Memory ran. */
export type NativeSyncState='unconfigured'|'idle'|'waiting'|'uploading'|'paused'|'blocked'|'error';
export type NativeErrorCode='connection_required'|'permission_required'|'retained_conflict'|'source_unavailable'|'transport_error'|'local_state_unavailable';
export type NativeStatusAction='configure_connection'|'grant_permission'|'resume_sync'|'retry_sync'|'open_processing';
export type NativeStatusFacts={
  pending:number|null; lastAcknowledgedAt?:string|null; archiveAcknowledged?:boolean;
  syncState:NativeSyncState; errorCode?:NativeErrorCode|null; retryAt?:string|null;
  /** The server owns processing state/actions; a successful upload never fills this field. */
  processing?:{state:string;allowedActions:string[];errorCode?:string|null;retryAt?:string|null}|null;
  scanComplete?:boolean|null; knownItems?:number|null; skipped?:number|null;
};
export type NativeStatusView={
  archive:{state:'unknown'|'local'|'partial'|'acknowledged';pending:number|null;lastAcknowledgedAt:string|null};
  processing:{state:string;allowedActions:string[];errorCode:string|null;retryAt:string|null};
  sync:{state:NativeSyncState;errorCode:NativeErrorCode|null;retryAt:string|null;allowedActions:NativeStatusAction[]};
  coverage:{state:'unknown'|'partial'|'scanned';knownItems:number|null;skipped:number|null};
};
export function nativeStatusView(facts:NativeStatusFacts):NativeStatusView{
  const count=(n:number|null|undefined)=>typeof n==='number'&&Number.isSafeInteger(n)&&n>=0?n:null;
  const pending=count(facts.pending),ack=facts.lastAcknowledgedAt||null,acknowledged=facts.archiveAcknowledged===true||Boolean(ack);
  const errorCode=facts.errorCode??(facts.syncState==='unconfigured'?'connection_required':facts.syncState==='error'?'transport_error':null);
  const allowedActions:NativeStatusAction[]=facts.syncState==='unconfigured'?['configure_connection']:
    errorCode==='permission_required'?['grant_permission']:
    facts.syncState==='paused'?['resume_sync']:
    facts.syncState==='uploading'?[]:['retry_sync'];
  if(facts.syncState!=='unconfigured'&&ack)allowedActions.push('open_processing');
  return {
    archive:{state:pending===null?'unknown':pending>0?(acknowledged?'partial':'local'):acknowledged?'acknowledged':'unknown',pending,lastAcknowledgedAt:ack},
    processing:{state:facts.processing?.state??'unknown',allowedActions:[...(facts.processing?.allowedActions??[])],errorCode:facts.processing?.errorCode??null,retryAt:facts.processing?.retryAt??null},
    sync:{state:facts.syncState,errorCode,retryAt:facts.retryAt??null,allowedActions},
    coverage:{state:facts.scanComplete==null?'unknown':facts.scanComplete?'scanned':'partial',knownItems:count(facts.knownItems),skipped:count(facts.skipped)},
  };
}

File: apps/android/app/src/main/java/dev/mote/collector/NativeStatus.kt
package dev.mote.collector

import org.json.JSONArray
import org.json.JSONObject

/** Mirrors @mote/shared/native-status; both implementations consume the same fixture. */
internal object NativeStatus {
    fun project(facts: JSONObject): JSONObject {
        fun count(key: String): Any = (facts.opt(key) as? Number)?.toDouble()?.takeIf { it.isFinite() && it >= 0 && it <= 9007199254740991.0 && it % 1 == 0.0 }?.toLong() ?: JSONObject.NULL
        fun optional(value: Any?): Any = value?.takeUnless { it == JSONObject.NULL } ?: JSONObject.NULL
        val pending = count("pending"); val ack = facts.optString("lastAcknowledgedAt").takeUnless { it.isBlank() || it == "null" }
        val acknowledged = facts.optBoolean("archiveAcknowledged") || ack != null
        val state = facts.getString("syncState")
        val error = facts.optString("errorCode").takeUnless { it.isBlank() || it == "null" } ?: when(state) { "unconfigured" -> "connection_required"; "error" -> "transport_error"; else -> null }
        val actions = JSONArray()
        when { state == "unconfigured" -> actions.put("configure_connection"); error == "permission_required" -> actions.put("grant_permission"); state == "paused" -> actions.put("resume_sync"); state != "uploading" -> actions.put("retry_sync") }
        if (state != "unconfigured" && ack != null) actions.put("open_processing")
        val processing = facts.optJSONObject("processing")
        return JSONObject()
            .put("archive", JSONObject().put("state", if (pending == JSONObject.NULL) "unknown" else if ((pending as Long) > 0) { if (!acknowledged) "local" else "partial" } else if (!acknowledged) "unknown" else "acknowledged").put("pending", pending).put("lastAcknowledgedAt", optional(ack)))
            .put("processing", JSONObject().put("state", processing?.optString("state") ?: "unknown").put("allowedActions", processing?.optJSONArray("allowedActions") ?: JSONArray()).put("errorCode", optional(processing?.opt("errorCode"))).put("retryAt", optional(processing?.opt("retryAt"))))
            .put("sync", JSONObject().put("state", state).put("errorCode", optional(error)).put("retryAt", optional(facts.opt("retryAt"))).put("allowedActions", actions))
            .put("coverage", JSONObject().put("state", if (!facts.has("scanComplete") || facts.isNull("scanComplete")) "unknown" else if (facts.getBoolean("scanComplete")) "scanned" else "partial").put("knownItems", count("knownItems")).put("skipped", count("skipped")))
    }
    fun summary(view: JSONObject): String = when(view.getJSONObject("archive").getString("state")) {
        "acknowledged" -> MoteI18n.text("当前待发已获中央确认")
        "partial" -> MoteI18n.text("部分资料已确认，仍有本机待发")
        "local" -> MoteI18n.text("资料保存在本机，等待中央确认")
        else -> MoteI18n.text("资料确认状态待查询")
    } + " · " + MoteI18n.text("中央处理状态待查询，请在中央任务中心查看")
}
